Fhi.Smittestopp.Verification.Domain/Factories/OneWayPseudonymFactory.cs

```diff
@@ -17,22 +16,30 @@ public OneWayPseudonymFactory(IOptions<Config> config)
 
         public string Create(string original)
         {
-            using (var hmacsha256 = new HMACSHA256(_config.KeyBytes))
-            {
-                var hash = hmacsha256.ComputeHash(Encoding.UTF8.GetBytes(original));
-                return Convert.ToBase64String(hash);
-            }
+            return Hash(original, _config.KeyBytes, KeyDerivationPrf.HMACSHA256, _config.NumIterations, _config.NumBytes);
```

Comment:
Sorry to bother again but I'd recommend not using a fixed salt.
By using `Pbkdf2` it's a lot better than before but it can still be improved with not that much work.

You could generate the salt with `RandomNumberGenerator`.
You then store the salt with the hash as `PasswordHasher` class is doing:
https://github.com/dotnet/aspnetcore/blob/9428ac6f4c2103227c46a4da23b4de9cb780dc08/src/Identity/Extensions.Core/src/PasswordHasher.cs#L153
And then when verifying the pseudonym just take the salt from hash.
https://github.com/dotnet/aspnetcore/blob/9428ac6f4c2103227c46a4da23b4de9cb780dc08/src/Identity/Extensions.Core/src/PasswordHasher.cs#L273

Reply:
Main benefit of unique salt is to get different hash with the same password input.
In our case it might not be relevant because I assume pseudonyms are all unique.

The one benefit I see is that when attacking the leaked hashes with some dictionary attacks it could make a difference.
With fixed salt you'd need to create only one dictionary (with this salt) to attack the hashes.
With unique salt you'd need to create a dictionary per each salt.

Reply:
I do certainly see your point about having a fixed salt making the process of attacking the hashes easier, but I see no practical way around this. Without a fixed salt, how do you suggest we match verification records from the same person?

If we generate a unique salt for each record, whenever we receive a new verification, does this not force us to hash the incoming ID with all existing salts in the dataset to find a match? I do not see how this is scalable.

Reply:
You're absolutely right.
It's as good as it can get!